### PR TITLE
Feature: Add LinearERC20VotingV1ValidatorV1

### PR DESCRIPTION
**Pull request by adamgall**
Depends on https://github.com/decentdao/decent-contracts/pull/254

Fixes [ENG-818](https://linear.app/decent-labs/issue/ENG-818/develop-validator-for-linearerc20votingv1)

**Summary:**

This PR introduces `LinearERC20VotingV1ValidatorV1`, a dedicated validator contract designed to work with `DecentPaymasterV1`. This validator specifically checks if a `vote(uint32, uint8)` UserOperation targeting a `LinearERC20VotingV1` strategy contract is likely to succeed before the paymaster agrees to sponsor it.

The PR includes the validator contract itself, a mock version of `LinearERC20VotingV1` for testing, and a comprehensive test suite for the new validator.

**Key Components:**

1.  **New Contract `LinearERC20VotingV1ValidatorV1.sol`:**
    *   Implements `IFunctionValidator`, `ERC165`, and `Version`.
    *   **`validateOperation(...)`:**
        *   Verifies the `callData` selector matches `ILinearERC20VotingV1.vote.selector`.
        *   Decodes `proposalId` and `voteType` from `callData`.
        *   Checks for valid `voteType` (0, 1, or 2).
        *   Retrieves proposal period (`startBlock`, `endBlock`) from the target `votingContract`.
        *   Ensures the proposal exists (non-zero `endBlock`).
        *   Verifies the voting period has not ended.
        *   Checks if the `lightAccountOwner` has already voted on the proposal.
        *   Retrieves the `governanceToken` from the `votingContract`.
        *   Fetches the `lightAccountOwner`'s checkpoints from the `governanceToken`.
        *   Returns `false` if the voter has no checkpoints.
        *   Iterates through checkpoints to find the voting weight at the `proposal.startBlock`.
        *   Returns `false` if the voting weight at `startBlock` is zero.
        *   Returns `true` if all checks pass, indicating the vote operation is likely to succeed.
    *   **`getVersion()`:** Returns the contract version (1).
    *   **`supportsInterface(...)`:** Implements ERC165 support for `IFunctionValidator`, `IVersion`, and `IERC165`.

2.  **New Mock Contract `MockLinearERC20VotingV1.sol`:**
    *   Provides a mock implementation of the functions from `LinearERC20VotingV1` and `IERC20Votes` that the validator interacts with.
    *   Allows setting proposal periods, voting ended status, voter status (`hasVoted`), governance token address, and checkpoint data for test scenarios.

3.  **New Test Suite `LinearERC20VotingV1ValidatorV1.test.ts`:**
    *   Provides comprehensive test coverage for `LinearERC20VotingV1ValidatorV1`.
    *   Tests various scenarios for `validateOperation`, including:
        *   Incorrect function selector in `callData`.
        *   Invalid vote types.
        *   Non-existent proposals.
        *   Voting period ended.
        *   User has already voted.
        *   Voter has zero voting weight (no checkpoints, all checkpoints after proposal start, relevant checkpoint has zero votes).
        *   Valid vote operations with various single and multiple checkpoint configurations.
        *   Edge cases like max voting weight and proposal start at block 0.
    *   Includes tests for ERC165 interface support and versioning.

**Motivation:**

*   To enable gasless voting for users interacting with `LinearERC20VotingV1` strategies via `DecentPaymasterV1`.
*   To provide a robust pre-validation mechanism that ensures the paymaster only sponsors vote operations that are likely to succeed, saving gas and preventing unnecessary sponsored transactions.
*   To create a reusable and upgradeable validator specific to the `LinearERC20VotingV1` contract's voting logic.

### CODE REVIEW

**Review by dacebt — COMMENTED** [review]
I'm stuck on these validators. Where are these used deployed/setup/used in the Gasless voting setup

**Review by dacebt — APPROVED** [review]
Thanks for the InG update, makes sense. Curious just had a thought this is seems very specific to voting and you said that only one validator will be assigned. Will this pattern allow for gasless options in the future?

**Comment by adamgall** [comment]
> Thanks for the InG update, makes sense. Curious just had a thought this is seems very specific to voting and you said that only one validator will be assigned. Will this pattern allow for gasless options in the future?

One validator is assigned _per target contract and function_. So yes, this pattern allows for additional transaction sponsorship (more than just voting) in the future.

File: contracts/deployables/account-abstraction/validators/LinearERC20VotingV1ValidatorV1.sol [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        // Iterate backwards through checkpoints to find the relevant one for startBlock.
+        // This is potentially more efficient than binary search if startBlock is recent.
+        uint256 votingWeight = 0;
+        for (uint256 i = numCheckpoints; i > 0; i--) {
+            // Checkpoint indices are 0-based, loop index 'i' is 1-based count.
+            Checkpoint memory checkpoint = governanceToken.checkpoints(
+                lightAccountOwner,
+                uint32(i - 1)
+            );
+
```
Comment (johnqh) [inline contracts/deployables/account-abstraction/validators/LinearERC20VotingV1ValidatorV1.sol:128]: I was thinking about this and there is another optimization we can do:
if (checkpoint.fromBlock > endBlock) {
// This means the vote has ended, since there is a checkpoint after the proposal endBlock. We can break out of the loop and return false.
}

**Review by johnqh — CHANGES_REQUESTED** [review]
Additional optimization we can do

File: contracts/deployables/account-abstraction/validators/LinearERC20VotingV1ValidatorV1.sol (hunk 1 above)
Comment (adamgall) [reply contracts/deployables/account-abstraction/validators/LinearERC20VotingV1ValidatorV1.sol:128]: updated

**Review by johnqh — APPROVED** [review]
